python/src/trezorlib/authentication.py

```diff
@@ -0,0 +1,345 @@
+        "TESTING ENVIRONMENT. DO NOT USE THIS DEVICE",
+        "Trezor Safe 3",
+        True,
+        _pk_p256(
+            "047f77368dea2d4d61e989f474a56723c3212dacf8a808d8795595ef38441427c"
+            "4389bc454f02089d7f08b873005e4c28d432468997871c0bf286fd3861e21e96a"
+        ),
+    ),
+    RootCertificate(
+        "TESTING ENVIRONMENT. DO NOT USE THIS DEVICE",
+        "Trezor Safe 5",
+        True,
+        _pk_p256(
+            "04e48b69cd7962068d3cca3bcc6b1747ef496c1e28b5529e34ad7295215ea161d"
+            "be8fb08ae0479568f9d2cb07630cb3e52f4af0692102da5873559e45e9fa72959"
+        ),
+    ),
+]
+
+
+class DeviceNotAuthentic(Exception):
+    pass
+
+
+class Certificate:
+    def __init__(self, cert_bytes: bytes) -> None:
+        self.cert_bytes = cert_bytes
+        self.cert = x509.load_der_x509_certificate(cert_bytes)
+
+    def __str__(self) -> str:
+        return self.cert.subject.rfc4514_string()
+
+    def public_key_bytes(self) -> bytes:
+        return self.cert.public_key().public_bytes(
+            serialization.Encoding.X962,
+            serialization.PublicFormat.UncompressedPoint,
+        )
+
+    def verify(self, signature: bytes, message: bytes) -> None:
+        cert_pubkey = self.cert.public_key()
+        assert isinstance(cert_pubkey, ec.EllipticCurvePublicKey)
+        cert_pubkey.verify(
+            self.fix_signature(signature),
+            message,
+            ec.ECDSA(hashes.SHA256()),
+        )
+
+    def verify_by(self, pubkey: ec.EllipticCurvePublicKey) -> None:
+        algo_params = self.cert.signature_algorithm_parameters
+        assert isinstance(algo_params, ec.ECDSA)
+        pubkey.verify(
+            self.fix_signature(self.cert.signature),
+            self.cert.tbs_certificate_bytes,
+            algo_params,
+        )
+
+    def is_issued_by(self, issuer: "Certificate", path_len: int) -> bool:
+        """Check if this certificate was issued by an issuer.
+
+        Returns True if:
+         * our `issuer` is the same as issuer's `subject`,
+         * the issuer is a valid CA, that is:
+           - has the cA flag set
+           - has a valid pathLenConstraint
+           - pathLenConstraint does not exceed the current path length.
+         * the issuer's public key signs this certificate.
+
+        Returns False if `issuer` is not permitted to issue this certificate, that is,
+        some of the above conditions are not met.
+
+        Raises InvalidSignature if `issuer` is valid but did not sign this certificate.
+        """
+        if issuer.cert.subject != self.cert.issuer:
+            LOG.error("Certificate %s is not issued by %s.", self, issuer)
+            return False
+
+        try:
+            # check issuer CA flag
+            ext = issuer.cert.extensions.get_extension_for_class(x509.BasicConstraints)
+            if ext.value.ca is not True:
+                LOG.error("Issuer is not a valid CA: %s", issuer)
+                return False
```

Comment:
The error handling here seems a bit inconsistent. This function indicates errors in two ways: returning `False` or raising an exception, which is a little hard to keep track of. Is there any reason for that?

Reply:
27b8f2703d21f43ceb6adf1b920683e4bcf79af5
i originally intended to differentiate between "bad signature" and "bad issuer" and this seemed like the easiest way to accomplish -- but switching to just bool return value actually makes the caller code nicer so :+1: